encode_decode.go

```diff
@@ -0,0 +1,455 @@
+						v.SetInt32Ptr(valGoField.Interface().(*int32))
+					case reflect.Int64:
+						v.SetInt64Ptr(valGoField.Interface().(*int64))
+					}
+				}
+				valTfField.Set(reflect.ValueOf(v))
+			case Float64Value:
+				switch valGoField.Kind() {
+				case reflect.Float32:
+					v.SetFloat32(valGoField.Interface().(float32))
+				case reflect.Float64:
+					v.SetFloat64(valGoField.Interface().(float64))
+				case reflect.Ptr:
+					switch valGoField.Type().Elem().Kind() {
+					case reflect.Float32:
+						v.SetFloat32Ptr(valGoField.Interface().(*float32))
+					case reflect.Float64:
+						v.SetFloat64Ptr(valGoField.Interface().(*float64))
+					}
+				}
+				valTfField.Set(reflect.ValueOf(v))
+			default:
+				// Default is not simple type. Require to get type in the goStruct to initialize the generic type in the tfStruct
+				// Define if is a slice, map or a struct in the goStruct
+				switch valGoField.Kind() {
+				case reflect.Map:
+
+					switch valGoField.Type().Elem().Kind() {
+					case reflect.String, reflect.Int64, reflect.Float64:
+						setOf(ctx, valTfField, valGoField)
+					case reflect.Ptr:
+
+						// Get the Tf value object by getOf method
+						TfV := getOf(ctx, valTfField)
+						for _, k := range valGoField.MapKeys() {
+							if !TfV.MapIndex(k).IsValid() {
+								TfV.SetMapIndex(k, reflect.New(TfV.Type().Elem().Elem()))
+							}
+
+							if err := encodeToTerraformStruct(ctx, TfV.MapIndex(k).Type(), TfV.MapIndex(k), valGoField.MapIndex(k).Type(), valGoField.MapIndex(k)); err != nil {
+								return err
+							}
+						}
+
+						setOf(ctx, valTfField, TfV)
+					}
+					// End of Map
+				case reflect.Slice:
+
+					switch valGoField.Type().Elem().Kind() {
+					case reflect.String, reflect.Int64, reflect.Float64:
+						setOf(ctx, valTfField, valGoField)
+					case reflect.Ptr:
+
+						// Get the Tf value object by getOf method
+						TfV := getOf(ctx, valTfField)
+						for j := 0; j < valGoField.Len(); j++ {
+							if !TfV.Index(j).IsValid() {
+								TfV.Index(j).Set(reflect.New(TfV.Type().Elem().Elem()))
+							}
+							if err := encodeToTerraformStruct(ctx, TfV.Index(j).Type(), TfV.Index(j), valGoField.Index(j).Type(), valGoField.Index(j)); err != nil {
+								return err
+							}
+						}
+
+						setOf(ctx, valTfField, TfV)
+					}
+					// End of Slice
+				case reflect.Ptr:
+
+					// Get the Tf value object by getOf method
+					TfV := getOf(ctx, valTfField)
+					if err := encodeToTerraformStruct(ctx, TfV.Type(), TfV, valGoField.Type(), valGoField); err != nil {
+						return err
+					}
+					setOf(ctx, valTfField, TfV)
+					// End of Ptr
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func decodeToStruct(ctx context.Context, rTfType reflect.Type, rTf reflect.Value, rGoType reflect.Type, rGo reflect.Value) error { //nolint:gocyclo
+	if rTf.Kind() == reflect.Ptr && rTf.IsNil() {
+		// TODO set rGo to nil
+		return nil
+	}
+
+	if rGo.Kind() == reflect.Ptr && rGo.IsNil() {
+		return nil
```

Comment:
return an error message

```suggestion
		return fmt.Errorf("%T canot be nil.", rGo.Interface(), rGoType)
```